### PR TITLE
Sdo/interactive

### PR DESCRIPTION
**Pull request by soerendomroes**
Added interactive options for layered and rectPacking.
Added interactive mode for rectPacking.

### CODE REVIEW

File: plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/GraphConfigurator.java [hunk 1]
```diff
@@ -213,6 +213,12 @@ public void prepareGraphForLayout(final LGraph lgraph) {
         if (lgraph.getProperty(LabelManagementOptions.LABEL_MANAGER) != null) {
             configuration.addAll(LABEL_MANAGEMENT_ADDITIONS);
         }
+        
+        // If the graph should be laid out interactively, add the layers and positions to the nodes.
+        if(lgraph.getProperty(LayeredOptions.INTERACTIVE_LAYOUT)) {
```
Comment (le-cds) [inline plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/GraphConfigurator.java]: Missing space after if.

File: plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/Layered.melk [hunk 2]
```diff
@@ -219,8 +224,27 @@ group layering {
...
-    
-    
+
+    advanced option layerChoiceConstraint: int {
+        label "Layer Choice Constraint"
+        description "Allows to set a constraint regarding the layer placement of a node. 
+                     Assumed the drawing has n layers and i < n.
+                     If set to i, it expresses that the node should be placed in i-th layer.
+                     Should i>=n be true then the node is placed in the last layer of the drawing.
+                    "
```
Comment (le-cds) [inline plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/Layered.melk]: Shouldn't be on a separate line.

File: plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/Layered.melk [hunk 3]
```diff
@@ -219,8 +224,27 @@ group layering {
         targets parents
         legacyIds de.cau.cs.kieler.klay.layered.wideNodesOnMultipleLayers
     }
-    
-    
+
+    advanced option layerChoiceConstraint: int {
+        label "Layer Choice Constraint"
+        description "Allows to set a constraint regarding the layer placement of a node. 
+                     Assumed the drawing has n layers and i < n.
```
Comment (le-cds) [inline plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/Layered.melk]: What is i?

File: plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/Layered.melk [hunk 4]
```diff
@@ -219,8 +224,27 @@ group layering {
...
+                     Should i>=n be true then the node is placed in the last layer of the drawing.
+                    "
+        default = -1
+        lowerBound = -1
+        targets nodes
+    }
+
+    output option layerID: int {
+        label "Layer ID"
+        description "Layer identifier that was calculated by Elk Layered for a node"
```
Comment (le-cds) [inline plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/Layered.melk]: ELK Layered

File: plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/ConstraintsPostprocessor.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+import org.eclipse.elk.alg.layered.graph.LNode.NodeType;
+import org.eclipse.elk.alg.layered.graph.Layer;
+import org.eclipse.elk.alg.layered.options.LayeredOptions;
+import org.eclipse.elk.core.alg.ILayoutProcessor;
+import org.eclipse.elk.core.util.IElkProgressMonitor;
+
+/**
+ * Adds to each LNode the layerID and positionID that has been computed by ELKLayered.
+ * 
+ * @author cos, jet
```
Comment (le-cds) [inline plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/ConstraintsPostprocessor.java]: ELK doesn't use author tags anymore.

File: plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/ConstraintsPostprocessor.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+ */
+public final class ConstraintsPostprocessor implements ILayoutProcessor<LGraph> {
+
+    /*
```
Comment (le-cds) [inline plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/ConstraintsPostprocessor.java]: Remove.

File: plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/ConstraintsPostprocessor.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+     * (non-Javadoc)
+     * 
+     * @see org.eclipse.elk.core.alg.ILayoutProcessor#process(java.lang.Object,
+     * org.eclipse.elk.core.util.IElkProgressMonitor)
+     */
+    @Override
+    public void process(LGraph graph, IElkProgressMonitor progressMonitor) {
```
Comment (le-cds) [inline plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/ConstraintsPostprocessor.java]: In ELK, parameters are supposed to be final.

File: plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/ConstraintsPostprocessor.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+        progressMonitor.begin("Constraints Postprocessor", 1);
+        progressMonitor.log("Constraints Postprocessor started");
+        progressMonitor.log(JsonDebugUtil.createDebugGraph(graph));
```
Comment (le-cds) [inline plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/ConstraintsPostprocessor.java]: There's a special log method for logging graphs, I believe. Also, the JSON representation should only be generated if logging is enabled in the first place.

File: plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/ConstraintsPostprocessor.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+                    currentNode.setProperty(LayeredOptions.LAYERING_LAYER_I_D, layerIndex);
+                    currentNode.setProperty(LayeredOptions.CROSSING_MINIMIZATION_POSITION_I_D, posIndex);
+                    posIndex++;
+                }
+            }
+
+            layerIndex++;
+        }
+        progressMonitor.log("Constraints Posprocessor should have set all ids");
+        progressMonitor.log(JsonDebugUtil.createDebugGraph(graph));
```
Comment (le-cds) [inline plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/ConstraintsPostprocessor.java]: See above.

File: plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/IntermediateProcessorStrategy.java [hunk 10]
```diff
@@ -328,6 +330,9 @@
 
         case SELF_LOOP_POSTPROCESSOR:
             return new SelfLoopPostProcessor();
+        
+        case CONSTRAINTS_POSTPROCESSOR:
```
Comment (le-cds) [inline plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/IntermediateProcessorStrategy.java]: We usually sort them alphabetically in this method.

File: plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/Layered.melk [hunk 11]
```diff
@@ -87,53 +87,53 @@ algorithm layered(LayeredLayoutProvider) {
...
-	supports org.eclipse.elk.separateConnectedComponents = true
-	supports org.eclipse.elk.insideSelfLoops.activate
-	supports org.eclipse.elk.insideSelfLoops.yo
-	supports org.eclipse.elk.nodeSize.constraints
-	supports org.eclipse.elk.nodeSize.options
-	supports org.eclipse.elk.direction = Direction.UNDEFINED
-	supports org.eclipse.elk.nodeLabels.placement
-	supports org.eclipse.elk.nodeLabels.padding
-	supports org.eclipse.elk.portLabels.placement
+    // other options
```
Comment (uruuru) [inline plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/Layered.melk]: I'm not sure if it's only due to these changes, but the indentation of the melk files is somewhat broken and inconsistent now. 
Can you please fix this?

File: plugins/org.eclipse.elk.alg.packing.rectangles/src/org/eclipse/elk/alg/packing/rectangles/RectPackingLayoutProvider.java [hunk 12]
```diff
@@ -60,13 +63,46 @@ public void layout(final ElkNode layoutGraph, final IElkProgressMonitor progress
...
+            EList<ElkNode> fixedNodes = new BasicEList<>();
+            for (ElkNode elkNode : rectangles) {
+                if (elkNode.hasProperty(RectPackingOptions.DESIRED_POSITION)) {
+                    fixedNodes.add(elkNode);
+                }
+            }
+            for (ElkNode elkNode : fixedNodes) {
+                rectangles.remove(elkNode);
+            }
+            ECollections.sort(fixedNodes, (a, b) -> {
```
Comment (uruuru) [inline plugins/org.eclipse.elk.alg.packing.rectangles/src/org/eclipse/elk/alg/packing/rectangles/RectPackingLayoutProvider.java]: Why not simply use `Collections.sort`? Avoid the EMF dependencies where possible within ELK.

File: plugins/org.eclipse.elk.alg.packing.rectangles/src/org/eclipse/elk/alg/packing/rectangles/RectPackingLayoutProvider.java [hunk 13]
```diff
@@ -60,13 +63,46 @@ public void layout(final ElkNode layoutGraph, final IElkProgressMonitor progress
...
         boolean expandToAspectRatio = layoutGraph.getProperty(RectPackingOptions.EXPAND_TO_ASPECT_RATIO);
+        // Whether interactive layout is activ.
+        boolean interactive = layoutGraph.getProperty(RectPackingOptions.INTERACTIVE);
 
         List<ElkNode> rectangles = layoutGraph.getChildren();
         DrawingUtil.resetCoordinates(rectangles);
         DrawingData drawing;
+        
+        if (interactive) {
+            EList<ElkNode> fixedNodes = new BasicEList<>();
```
Comment (uruuru) [inline plugins/org.eclipse.elk.alg.packing.rectangles/src/org/eclipse/elk/alg/packing/rectangles/RectPackingLayoutProvider.java]: Why do you require an EMF list here?

File: plugins/org.eclipse.elk.alg.packing.rectangles/src/org/eclipse/elk/alg/packing/rectangles/RectPackingLayoutProvider.java (hunk 12 above)
Comment (soerendomroes) [reply plugins/org.eclipse.elk.alg.packing.rectangles/src/org/eclipse/elk/alg/packing/rectangles/RectPackingLayoutProvider.java]: You are right, I will correct this.

File: plugins/org.eclipse.elk.alg.packing.rectangles/src/org/eclipse/elk/alg/packing/rectangles/RectPackingLayoutProvider.java (hunk 13 above)
Comment (soerendomroes) [reply plugins/org.eclipse.elk.alg.packing.rectangles/src/org/eclipse/elk/alg/packing/rectangles/RectPackingLayoutProvider.java]: See above.

**Pull request merged**